src/EnergyPlus/api/datatransfer.h

```diff
@@ -121,6 +121,12 @@ ENERGYPLUSLIB_API int apiErrorFlag(EnergyPlusState state);
 ///          a calculation to continue, this function can reset the flag.
 /// \param[in] state An active EnergyPlusState instance created with `stateNew`.
 ENERGYPLUSLIB_API void resetErrorFlag(EnergyPlusState state);
+/// \brief Provides the input file path back to the user
+/// \details In most circumstances the client will know the path to the input file, but there are some cases where code
+///          is generalized in unexpected workflows.  Users have requested a way to get the input file path back from the running instance.
+/// \param[in] state An active EnergyPlusState instance created with `stateNew`.
+/// \return A char * of the input file path.  This allocates a new char *, and calling clients must free this when done with it!
+ENERGYPLUSLIB_API char *inputFilePath(EnergyPlusState state);
```

Comment:
User is warned this allocates, good

Reply:
Yes....but now I think I need to add something to the Python API.  Once Python calls this function and converts it into a Python string (PyObject*), the C string is still hanging out.  I may need a very minimal function in the C API that can free one of these char*.  Thoughts @jmarrec ?

Reply:
I think you can to decode it to utf-8, then just free it, right?

Reply:
Same as https://github.com/NREL/EnergyPlus/blob/52fee4b00a1d679d545dfaf69a3704227b11ff4c/src/EnergyPlus/api/datatransfer.py#L374-L375

Reply:
Speaking off, you should:

* Use FileSystem::toGenericString so you don't have issues with special chars
* Probably have them return a pathlib.Path instead.

Reply:
Done in 2c31f9fc35b78f28875f5545f59adc17bc5bd43b.


And you don't need to free the c_char_p returned by the C fun: ,if you set restype to `c_char_p`, ctypes returns a regular Python bytes object. and it seems it manages the memory for you. I tested with and without a C delete[] call, with ASAN enabled, and it works correctly when you do nothing, and it gives a malloc error if you try to free.


```
> /Users/julien/Software/Others/EnergyPlus-build/Products/pyenergyplus/datatransfer.py(377)get_input_file_path()
-> c_string = self.api.inputFilePath(state)
(Pdb) n
> /Users/julien/Software/Others/EnergyPlus-build/Products/pyenergyplus/datatransfer.py(378)get_input_file_path()
-> res = Path(c_string.decode('utf-8'))
(Pdb) p c_srtring
*** NameError: name 'c_srtring' is not defined
(Pdb) p c_string
b'/Users/julien/Software/Others/EnergyPlus/testfiles/_1ZoneUncontrolled_ForAPITesting.idf'
(Pdb) type(c_string)
<class 'bytes'>
(Pdb) n
> /Users/julien/Software/Others/EnergyPlus-build/Products/pyenergyplus/datatransfer.py(379)get_input_file_path()
-> self.api.freeCString(c_string)
(Pdb) s
=================================================================
==47091==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x0001072b28d0 in thread T0
    #0 0x1016f5c88 in wrap__ZdaPv+0x74 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x61c88)
    #1 0x116cd5a04 in freeCString datatransfer.cc:293
    #2 0x19b79504c in ffi_call_SYSV+0x4c (libffi.dylib:arm64e+0x804c)
    #3 0x19b79dadc in ffi_call_int+0x4b8 (libffi.dylib:arm64e+0x10adc)
    #4 0x1068b3ef4 in _ctypes_callproc+0x530 (_ctypes.cpython-39-darwin.so:arm64+0xbef4)
    #5 0x1068ad238 in PyCFuncPtr_call+0x48c (_ctypes.cpython-39-darwin.so:arm64+0x5238)
    #6 0x1021d3584 in _PyObject_MakeTpCall+0x164 (libpython3.9.dylib:arm64+0x3f584)
    #7 0x1022a9620 in call_function+0x1fc (libpython3.9.dylib:arm64+0x115620)
    #8 0x1022a6c38 in _PyEval_EvalFrameDefault+0x5a24 (libpython3.9.dylib:arm64+0x112c38)
    #9 0x1021d3d7c in function_code_fastcall+0x6c (libpython3.9.dylib:arm64+0x3fd7c)
    #10 0x1022a95c0 in call_function+0x19c (libpython3.9.dylib:arm64+0x1155c0)
    #11 0x1022a6c14 in _PyEval_EvalFrameDefault+0x5a00 (libpython3.9.dylib:arm64+0x112c14)
    #12 0x1021d3d7c in function_code_fastcall+0x6c (libpython3.9.dylib:arm64+0x3fd7c)
    #13 0x1068b31b4 in closure_fcn+0x21c (_ctypes.cpython-39-darwin.so:arm64+0xb1b4)
    #14 0x19b79df2c in ffi_closure_SYSV_inner+0x330 (libffi.dylib:arm64e+0x10f2c)
    #15 0x19b7951e4 in ffi_closure_SYSV+0x34 (libffi.dylib:arm64e+0x81e4)
    #16 0x116c99d68 in decltype(std::declval<void (*&)(void*)>()(std::declval<void*>())) std::__1::__invoke[abi:ue170006]<void (*&)(void*), void*>(void (*&)(void*), void*&&) invoke.h:340
    #17 0x116c99c94 in void std::__1::__invoke_void_return_wrapper<void, true>::__call[abi:ue170006]<void (*&)(void*), void*>(void (*&)(void*), void*&&) invoke.h:415
    #18 0x116c99c68 in std::__1::__function::__alloc_func<void (*)(void*), std::__1::allocator<void (*)(void*)>, void (void*)>::operator()[abi:ue170006](void*&&) function.h:193
    #19 0x116c95a44 in std::__1::__function::__func<void (*)(void*), std::__1::allocator<void (*)(void*)>, void (void*)>::operator()(void*&&) function.h:364
    #20 0x116c77aa4 in std::__1::__function::__value_func<void (void*)>::operator()[abi:ue170006](void*&&) const function.h:518
    #21 0x116c778a8 in std::__1::function<void (void*)>::operator()(void*) const function.h:1169
    #22 0x11af8c4f8 in EnergyPlus::PluginManagement::runAnyRegisteredCallbacks(EnergyPlus::EnergyPlusData&, EnergyPlus::EMSManager::EMSCallFrom, bool&) PluginManager.cc:142
    #23 0x1184ecd78 in EnergyPlus::EMSManager::ManageEMS(EnergyPlus::EnergyPlusData&, EnergyPlus::EMSManager::EMSCallFrom, bool&, ObjexxFCL::Optional<int const>) EMSManager.cc:276
    #24 0x119816ab8 in EnergyPlus::HeatBalanceManager::ManageHeatBalance(EnergyPlus::EnergyPlusData&) HeatBalanceManager.cc:218
    #25 0x11b98a644 in EnergyPlus::SimulationManager::ManageSimulation(EnergyPlus::EnergyPlusData&) SimulationManager.cc:497
    #26 0x116d1b794 in runEnergyPlusAsLibrary(EnergyPlus::EnergyPlusData&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) EnergyPlusPgm.cc:458
    #27 0x116c7f924 in energyplus runtime.cc:76
    #28 0x19b79504c in ffi_call_SYSV+0x4c (libffi.dylib:arm64e+0x804c)
    #29 0x19b79dadc in ffi_call_int+0x4b8 (libffi.dylib:arm64e+0x10adc)
    #30 0x1068b3ef4 in _ctypes_callproc+0x530 (_ctypes.cpython-39-darwin.so:arm64+0xbef4)
    #31 0x1068ad238 in PyCFuncPtr_call+0x48c (_ctypes.cpython-39-darwin.so:arm64+0x5238)
    #32 0x1021d3584 in _PyObject_MakeTpCall+0x164 (libpython3.9.dylib:arm64+0x3f584)
    #33 0x1022a9620 in call_function+0x1fc (libpython3.9.dylib:arm64+0x115620)
    #34 0x1022a6c38 in _PyEval_EvalFrameDefault+0x5a24 (libpython3.9.dylib:arm64+0x112c38)
    #35 0x1021d3d7c in function_code_fastcall+0x6c (libpython3.9.dylib:arm64+0x3fd7c)
    #36 0x1022a95c0 in call_function+0x19c (libpython3.9.dylib:arm64+0x1155c0)
    #37 0x1022a6c14 in _PyEval_EvalFrameDefault+0x5a00 (libpython3.9.dylib:arm64+0x112c14)
    #38 0x1022aa4e8 in _PyEval_EvalCode+0xba8 (libpython3.9.dylib:arm64+0x1164e8)
    #39 0x1022a1148 in PyEval_EvalCode+0x4c (libpython3.9.dylib:arm64+0x10d148)
    #40 0x1022e8f98 in pyrun_file+0x12c (libpython3.9.dylib:arm64+0x154f98)
    #41 0x1022e6f6c in PyRun_SimpleFileExFlags+0x26c (libpython3.9.dylib:arm64+0x152f6c)
    #42 0x102303f40 in Py_RunMain+0x6a0 (libpython3.9.dylib:arm64+0x16ff40)
    #43 0x102304404 in pymain_main+0x150 (libpython3.9.dylib:arm64+0x170404)
    #44 0x102304480 in Py_BytesMain+0x24 (libpython3.9.dylib:arm64+0x170480)
    #45 0x18a4460dc  (<unknown module>)

Address 0x0001072b28d0 is a wild pointer inside of access range of size 0x000000000001.
SUMMARY: AddressSanitizer: bad-free (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x61c88) in wrap__ZdaPv+0x74
==47091==ABORTING
Abort trap: 6
```